File: ios/CognitiveDissonance/BabylonNativeViewManager.swift
//
//  BabylonNativeViewManager.swift
//  CognitiveDissonance
//
//  Babylon Native View Manager for iOS
//  Bridges MTKView → Babylon Native Engine (Metal)
//
//  This is a STUB implementation. Full implementation requires:
//  1. Babylon Native iOS framework integration (Metal backend)
//  2. MTKView setup with Metal device and command queue
//  3. Babylon Native engine initialization with Metal surface
//  4. Engine reference bridging to JavaScript via RCTEventEmitter
//
//  See design.md "Babylon Native Integration Architecture" for full spec.
//

import Foundation
import UIKit
import React

// MARK: - Placeholder View with exported properties

/// Custom UIView subclass that holds the properties exported via RCT_EXPORT_VIEW_PROPERTY
/// in BabylonNativeViewManager.m. React Native sets these properties directly on the view
/// returned by the view manager's view() method.
@objc class BabylonNativeView: UIView {
  @objc var onEngineReady: RCTDirectEventBlock?
  @objc var antialias: Bool = true
  @objc var stencil: Bool = true
}

@objc(BabylonNativeViewManager)
class BabylonNativeViewManager: RCTViewManager {

  override func view() -> UIView! {
    // TODO: Replace with MTKView + Babylon Native Metal engine
    // For now, return a placeholder view with error message
    let placeholderView = BabylonNativeView()
    placeholderView.backgroundColor = UIColor.black

    let label = UILabel()
    label.text = "Babylon Native not implemented\nFalling back to screen mode"
    label.textColor = UIColor.white
    label.textAlignment = .center
    label.numberOfLines = 0
    label.translatesAutoresizingMaskIntoConstraints = false

    placeholderView.addSubview(label)

    NSLayoutConstraint.activate([
      label.centerXAnchor.constraint(equalTo: placeholderView.centerXAnchor),
      label.centerYAnchor.constraint(equalTo: placeholderView.centerYAnchor),
      label.leadingAnchor.constraint(equalTo: placeholderView.leadingAnchor, constant: 20),
      label.trailingAnchor.constraint(equalTo: placeholderView.trailingAnchor, constant: -20)
    ])

    return placeholderView
  }

  override static func requiresMainQueueSetup() -> Bool {
    return true
  }
}

// MARK: - Full Implementation Outline
//
// 1. Import Babylon Native framework:
//    import BabylonNative
//
// 2. Create MTKView with Metal device:
//    let metalDevice = MTLCreateSystemDefaultDevice()
//    let mtkView = MTKView(frame: .zero, device: metalDevice)
//    mtkView.colorPixelFormat = .bgra8Unorm
//    mtkView.depthStencilPixelFormat = .depth32Float_stencil8
//
// 3. Initialize Babylon Native engine:
//    let engine = BabylonNative.Engine(metalView: mtkView)
//
// 4. Bridge engine reference to JS:
//    let engineId = UUID().uuidString
//    // Store engine in global registry keyed by engineId
//    // Send onEngineReady event with engineId
//
// 5. Implement render loop:
//    mtkView.delegate = self
//    func draw(in view: MTKView) {
//      engine.renderFrame()
//    }
//
// See Babylon Native iOS samples for reference implementation:
// https://github.com/BabylonJS/BabylonNative/tree/master/Apps/Playground/iOS

File: android/app/src/main/java/arcade/cabinet/cognitivedissonance/BabylonNativeViewManager.kt
/**
 * BabylonNativeViewManager.kt
 * CognitiveDissonance
 *
 * Babylon Native View Manager for Android
 * Bridges SurfaceView → Babylon Native Engine (Vulkan/GLES)
 *
 * This is a STUB implementation. Full implementation requires:
 * 1. Babylon Native Android library integration (Vulkan/GLES backend)
 * 2. SurfaceView setup with EGL context
 * 3. Babylon Native engine initialization with native surface
 * 4. Engine reference bridging to JavaScript via ReactContextBaseJavaModule
 *
 * See design.md "Babylon Native Integration Architecture" for full spec.
 */

package arcade.cabinet.cognitivedissonance

import android.graphics.Color
import android.view.Gravity
import android.view.View
import android.widget.TextView
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.uimanager.SimpleViewManager
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.annotations.ReactProp

class BabylonNativeViewManager(
    private val reactContext: ReactApplicationContext
) : SimpleViewManager<View>() {

    override fun getName() = "BabylonNativeView"

    override fun createViewInstance(reactContext: ThemedReactContext): View {
        // TODO: Replace with SurfaceView + Babylon Native Vulkan/GLES engine
        // For now, return a placeholder view with error message
        val textView = TextView(reactContext)
        textView.text = "Babylon Native not implemented\nFalling back to screen mode"
        textView.setTextColor(Color.WHITE)
        textView.setBackgroundColor(Color.BLACK)
        textView.gravity = Gravity.CENTER
        textView.textSize = 16f
        return textView
    }

    @ReactProp(name = "antialias")
    fun setAntialias(view: View, antialias: Boolean) {
        // TODO: Configure MSAA on SurfaceView
    }

    @ReactProp(name = "stencil")
    fun setStencil(view: View, stencil: Boolean) {
        // TODO: Configure stencil buffer on EGL context
    }
}

// MARK: - Full Implementation Outline
//
// 1. Import Babylon Native library:
//    import com.babylonjs.BabylonNative
//
// 2. Create SurfaceView with EGL context:
//    val surfaceView = SurfaceView(reactContext)
//    val eglContext = EGL14.eglCreateContext(...)
//    surfaceView.holder.addCallback(object : SurfaceHolder.Callback {
//      override fun surfaceCreated(holder: SurfaceHolder) {
//        // Initialize Babylon Native engine with native surface
//      }
//    })
//
// 3. Initialize Babylon Native engine:
//    val engine = BabylonNative.Engine(surfaceView.holder.surface)
//
// 4. Bridge engine reference to JS:
//    val engineId = UUID.randomUUID().toString()
//    // Store engine in global registry keyed by engineId
//    // Send onEngineReady event with engineId via RCTEventEmitter
//
// 5. Implement render loop:
//    val renderThread = Thread {
//      while (running) {
//        engine.renderFrame()
//      }
//    }
//    renderThread.start()
//
// See Babylon Native Android samples for reference implementation:
// https://github.com/BabylonJS/BabylonNative/tree/master/Apps/Playground/Android
